### PR TITLE
Feature: BB-213 fix oplog populator race condition

### PR DESCRIPTION
**Pull request by welansari**
A race condition occurs when multiple events are processed at the same time
and that affect the same connector, in which case the connector is not configured
correctly.

For example when two buckets are being added to the same connector at the same time,
only one of them gets added, this is caused by the two operations getting and updating
the list of buckets assigned to the connector at the same time, which results in one of
them overwriting the other.

The solution is using a scheduler that periodically updates all of the connectors.
Connectors are no longer updated on an event basis

### CODE REVIEW

File: extensions/oplogPopulator/OplogPopulatorConfigValidator.js
Comment (francoisferrand) [inline extensions/oplogPopulator/OplogPopulatorConfigValidator.js]: every 5 minutes is a long time : we may miss lots of events  :-/

From what i understand, the problem is that we get the list of connectors every time : just getting the list once on startup (before listening to mongo events) should fix the problem as well, without introducing 5 minute delay?

Then, we may introduce a _short_ delay, to avoid too many reconfig : like once we get an update, wait a few seconds before applying it (so that we apply everything that happened since)...

Comment (welansari) [reply extensions/oplogPopulator/OplogPopulatorConfigValidator.js]: I actually used the same value used for all backbeat recurring tasks 5 seconds.

The issue is not in getting the list of connectors but the list of buckets assigned to a connector, this can not be done once as the values change with each update.

So with each new bucket to add (or remove) from a connector we have to get that list of buckets and update it.

When having multiple updates at the same time, the update tasks might get the wrong list (not updated), add the task's bucket and then configure the connector. This can lead to not taking in consideration the other updates.

File: extensions/oplogPopulator/modules/ConnectorsManager.js
Comment (jbertran) [inline extensions/oplogPopulator/modules/ConnectorsManager.js]: This does not have to be inside the `try` block.

File: extensions/oplogPopulator/modules/ConnectorsManager.js
Comment (jbertran) [inline extensions/oplogPopulator/modules/ConnectorsManager.js]: This output is a little bit misleading despite the dissimilar log levels.

Can we change the message according to whether all connector updates succeeded, or connectors partially succeeded?

**Review by jbertran — CHANGES_REQUESTED** [review]

**Review by jbertran — APPROVED** [review]

File: extensions/oplogPopulator/modules/ConnectorsManager.js
Comment (ghost) [inline extensions/oplogPopulator/modules/ConnectorsManager.js]: Nit: I think it would make sense to log as `error` if all updates failed (allowing to dump the debug logs as well if an error occurred), a `warn` if some of them failed and an `info` log it it went well.

**Review by ghost — COMMENTED** [review]
If I understand correctly, while I understand that removing the event-based update of the connector bucket list will help us dealing with concurrent updates/race conditions, I am not 100% sure if it is acceptable to "ignore" changes every 5s for 5s as it may lead to data loss or unwanted writes, if we're unlucky (and we will, in production!).

So I was wondering if a _stream_ approach would help us keeping the events in the right order, as they arrive, and avoid concurrent requests overriding each other? Basically, anything that could converge towards a (100%) safe solution, is this possible?

**Comment by welansari** [comment]
@williamlardier At a second thought i do agree that even 5 seconds, is a long time for the kind of updates we are doing.
A stream based system won't solve the concurrency issue as we are still dealing with async event handlers, i believe we need controlled connector updates.

What i did for now is reduce the delay between updates to one second, but this time we only update the connectors when there are actual changes made to them (buckets added or removed), this should help reduce network calls.

File: extensions/oplogPopulator/modules/Connector.js
Comment (ghost) [inline extensions/oplogPopulator/modules/Connector.js:130]: What happen if the `updatePipeline` fails? do we still add the bucket to the list, as it is done before the try/catch?
(I'm curious to know if the approach will be stable in case of errors, so we don't end up in an inconsistent state, but I may lack a bit of context on this one)

File: extensions/oplogPopulator/modules/Connector.js
Comment (ghost) [inline extensions/oplogPopulator/modules/Connector.js]: Nit:
```suggestion
    _updateConnectorState(bucketsGotModified, timeBeforeUpdate = 0) {
```

to be consistent with the jsdoc type

**Review by ghost — COMMENTED** [review]
Thanks for the update, I think it's a better first step until controlled connector updates, maybe. Please find one question and one suggested change before I approve

File: extensions/oplogPopulator/modules/Connector.js
Comment (welansari) [reply extensions/oplogPopulator/modules/Connector.js:130]: Yes the bucket is still added to the list but `_state.bucketsGotModified` stays as true, which will trigger the next update of the connector, so we will basically be retrying until it succeeds. We also get an error log each time a connector fails

Comment (welansari) [reply extensions/oplogPopulator/modules/Connector.js:130]: i added this https://github.com/scality/backbeat/compare/17128feaebde3bc87cd6fe8afe642ba5063f0702..a0941c5e9287ea76513da7b7e7104476a3c38bc9 just to be sure the retry gets triggered

**Review by ghost — APPROVED** [review]
LGTM, but might be better to wait for senior reviews on this one :slightly_smiling_face:

**Pull request merged**
